server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -285,23 +286,65 @@ static void resolvePipelinesAndUpdateIndexRequest(
         if (indexRequest.isPipelineResolved()) {
             return;
         }
+        var pipelines = resolveStoredPipelines(originalRequest, indexRequest, metadata, epochMillis);
+        setPipelineOnRequest(indexRequest, pipelines);
+    }
 
-        /*
-         * Here we look for the pipelines associated with the index if the index exists. If the index does not exist we fall back to using
-         * templates to find the pipelines.
-         */
-        final Pipelines pipelines = resolvePipelinesFromMetadata(originalRequest, indexRequest, metadata, epochMillis).or(
-            () -> resolvePipelinesFromIndexTemplates(indexRequest, metadata)
-        ).orElse(Pipelines.NO_PIPELINES_DEFINED);
+    public static boolean resolveAndUpdateAllPipelines(List<DocWriteRequest<?>> requests, Metadata metadata) {
+        final Map<String, IngestService.Pipelines> storedPipelineCache = new HashMap<>();
+        boolean hasIndexRequestsWithPipelines = false;
+        for (DocWriteRequest<?> actionRequest : requests) {
+            IndexRequest indexRequest = TransportAbstractBulkAction.getIndexWriteRequest(actionRequest);
+            if (indexRequest != null) {
+                if (indexRequest.isPipelineResolved() == false) {
+                    // Resolve the pipeline from setting or templates if not cached
+                    var pipelines = storedPipelineCache.computeIfAbsent(
+                        indexRequest.index(),
```

Comment:
This is now only looking up pipelines per index (rather than per index request) for all requests. So this will apply whether or not lazy rollover is set.